### PR TITLE
[ExpressionLanguage] Fix matches when the regexp is not valid

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | n/a 
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Symfony/Component/ExpressionLanguage/Tests/Node/BinaryNodeTest.php [hunk 1]
```diff
@@ -111,7 +114,7 @@ public function getCompileData()
 
             ['range(1, 3)', new BinaryNode('..', new ConstantNode(1), new ConstantNode(3))],
 
-            ['preg_match("/^[a-z]+/i\$/", "abc")', new BinaryNode('matches', new ConstantNode('abc'), new ConstantNode('/^[a-z]+/i$/'))],
+            ['(false !== $ret = @preg_match("/^[a-z]+\$/", "abc") ? $ret : throw new SyntaxException(sprintf(\'Regexp "%s" passed to "matches" is not valid.\'),"/^[a-z]+\$/")))', new BinaryNode('matches', new ConstantNode('abc'), new ConstantNode('/^[a-z]+$/'))],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Tests/Node/BinaryNodeTest.php]: syntax for "throw" won't work on php < 8
this one should:
`(static function () { throw new SyntaxException...})()`

Comment (nicolas-grekas) [reply src/Symfony/Component/ExpressionLanguage/Tests/Node/BinaryNodeTest.php]: or even better!
`(static function ($r, $s) { set_error_handler(function ($t, $m) use ($r) { throw new SyntaxException(sprintf('Regexp "%s" passed to "matches" is not valid', $r).substr($m, 12)); }); try { return preg_match($r,
$s); } finally { restore_error_handler(); } })($r, $s)`

Comment (fabpot) [reply src/Symfony/Component/ExpressionLanguage/Tests/Node/BinaryNodeTest.php]: I've done the first one to keep things simple.

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/ExpressionLanguage/Tests/Node/BinaryNodeTest.php [hunk 2]
```diff
@@ -160,7 +163,42 @@ public function getDumpData()
...
+    public function testCompileMatchesWithInvalidRegexp()
+    {
+        $node = new BinaryNode('matches', new ConstantNode('abc'), new ConstantNode('this is not a regexp'));
+
+        $this->expectExceptionObject(new SyntaxError('Regexp "this is not a regexp" passed to "matches" is not valid: Delimiter must not be alphanumeric or backslash'));
+        $compiler = new Compiler([]);
+        $node->compile($compiler);
+    }
+
+    public function testCompileMatchesWithInvalidRegexpAsExpression()
```
Comment (tantraluhur) [inline src/Symfony/Component/ExpressionLanguage/Tests/Node/BinaryNodeTest.php:195]: mantap bro

**Pull request merged**